src/sbml/conversion/SBMLRateRuleConverter.cpp

```diff
@@ -0,0 +1,1090 @@
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Species* species = model->getSpecies(node->getName());
+    bool isSpecies = (species != NULL);
+    bool isVariable = !node->isConstant();
+    return isVariable && isSpecies;
+}
+
+bool SBMLRateRuleConverter::isNumericalConstantOrParameter(ASTNode* node, Model* model)
+{
+    if (!node->isName()) // some nodes, like * operators, don't seem to have a name in the first place
+        return false;
+    Parameter* parameter = model->getParameter(node->getName());
+    bool isParameter = (parameter != NULL);
+    bool isNumericalConstant = node->isNumber() && node->isConstant();
+    return isNumericalConstant || isParameter;
+}
+
+void
+SBMLRateRuleConverter::dealWithSpecies()
+{
+  // loop through ODE pairs and check that any variables are listed as Species
+  for (odeIt it = mODEs.begin(); it != mODEs.end(); ++it)
+  {
+    const std::string variable = it->first;
+    // already a species
+    if (mDocument->getModel()->getSpecies(variable) != NULL)
+    {
+      continue;
+    }
+    else
+    {
+      // must be a parameter
+      Parameter *p = mDocument->getModel()->removeParameter(variable);
+      Species *newSpecies = mDocument->getModel()->createSpecies();
+      newSpecies->setId(variable);
+      newSpecies->setInitialAmount(p->getValue());
+      newSpecies->setHasOnlySubstanceUnits(true);
+      newSpecies->setBoundaryCondition(false);
+      newSpecies->setConstant(false);
+      if (mDocument->getModel()->getCompartment(0) == NULL)
+      {
+        Compartment * newCompartment = mDocument->getModel()->createCompartment();
+        newCompartment->setId("addedCompartment");
+        newCompartment->setConstant(true);
+        newCompartment->setSize(1.0);
+        newSpecies->setCompartment("addedCompartment");
+      }
+      else
+      {
+        newSpecies->setCompartment(mDocument->getModel()->getCompartment(0)->getId());
+      }
+      delete p;
+    }
+  }
+}
+void
+SBMLRateRuleConverter::createReactions()
+{
+  unsigned int i = 0;
+  for (setCoeffIt it = mCoefficients.begin(); it != mCoefficients.end(); ++it)
+  {
+    Reaction *r = mDocument->getModel()->createReaction();
+    r->setReversible(false);
+    r->setFast(false);
+    for (unsigned int j = 0; j < mODEs.size(); ++j)
+    {
+      double stoichiometry = 1.0;
+      if (mReactants[i][j] > 0)
+      {
+        stoichiometry = mReactants[i][j];
+        SpeciesReference *sr = r->createReactant();
+        sr->setSpecies(mODEs[j].first);
+        sr->setStoichiometry(stoichiometry);
+        sr->setConstant(true);
+      }
+      if (mProducts[i][j] > 0)
+      {
+        stoichiometry = mProducts[i][j];
+        SpeciesReference *sr = r->createProduct();
+        sr->setSpecies(mODEs[j].first);
+        sr->setStoichiometry(stoichiometry);
+        sr->setConstant(true);
+      }
+      if (mModifiers[i][j] > 0)
```

Comment:
currently this leads to many modifiers being created if a symbol is used multiple times), please change to: 

`      if (mModifiers[i][j] > 0 && r->getModifier(mODEs[j].first) == NULL)`